daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SeparatedSequenceParsers.scala

```diff
@@ -299,6 +299,18 @@ final class OrderedSeparatedSequenceParser(
 
       val bitPosBeforeSeparator = pstate.bitPos0b
 
+      pstate.pushDiscriminator
```

Comment:
If we pushing the discriminator here, it means that if a separator parser fails and processFailedSeparatorsWithPoU is called, the condition that checks pstate.discriminator == true will never pass and Failed_WithDiscriminatorSet will never be used (at least because of a missing separator), I think? So it seems like that condition is never going to be hit?

If so, maybe the right fix is to put the pushDiscriminator back where it was, and just remove the pstate.discriminitor == true check so that processFailedSeparatorsWithPoU always returns Failed_SpeculativeParse if a separator parser fails? Or is there some more complexity related to separators and discriminators that is missing?

Reply:
The pushD gives a clean false dis criminaltor at start. The only way that disc is set is when we parse The element Itself - which happens after the infix sep as you pointed out - so there can be a disc set after a postfix sep fails but not prefix or infix.
I am considering moving the push/ pop into the base class method so it will be centralized and not even an issue here.